### DISCUSSION TITLE
Possible security issue with http upgrade and X-Forwarded-Host header

### DISCUSSION THREAD

**Discussion by lodrantj** [Kong/kong#9454] [Help]
Hi, everybody

A security expert ran a series of penetration test on our cloud system which uses kong as a reverse proxy. One reported issue (not marked as extremely severe, but puzzling nonetheless) is about a possible man-in-the-middle attack when someone is able to add an X-Forwarded-Host header into a clients http request.

Our kong is setup for http upgrade on every request, responding to http reuqests with a HTTP 301 reponse, sending a https based address to the client in the Location response header, which is what is expected.
What the expert noticed is that, when the initial request (http://) contains the X-Forwarded-Host header with any value, the 301 response from kong includes the Location header pointing to https:// + whatever is the value of X-Forwarded-Host header.

The effectively means the client is redirected to an arbitrary site based on the header's value. I tried several steps using response transformer and even pre/post-functions modifying ngx.var.upstream_x_forwarded_* values, but nothing helped. Seems like the 301 redirection for http upgrade happens before any other processing of the request takes place.

Is my analysis correct? Is there anything I can do about it?
Ideally I would like to be able to either ignore the X-Forwarded_Host value from the client (I do not want to be proxied) or at least only allow redirection to specific known domains.

I hope someone is able to shed some light on this issue...

**Comment by StarlightIbuki** [comment]
Hi. Are you using some plugin or some custom plugin to redirect the request? Or are you using the header to route?
Because if a route's protocol is limited to `https`, it responds `http` with 426 (not 301), and if we use `request_termination` to return a 301, there's no such behavior.

**Comment by hbagdi** [comment]
This is a fair issue. Kong should take x-forwarded-host into account only when the request comes in from trusted_ips. If there isn't another proxy in front of Kong, then Kong should ignore x-forwarded-host header altogether.
